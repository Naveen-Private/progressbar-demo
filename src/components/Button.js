import React from 'react'
import styled from 'styled-components'

const ButtonContainer = styled.button`
    padding: 12px 24px;
    border-radius: 5px;
    margin-left: 18px;
    margin-right: 18px;
    background: transparent;
`

export default class Button extends React.Component {
    render(){
        const {
            btnText,
            updateProgressBar
        } = this.props
        return(
            <ButtonContainer onClick={updateProgressBar}>{btnText}</ButtonContainer>
        )
    }
}