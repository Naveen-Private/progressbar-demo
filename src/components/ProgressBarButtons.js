import React from 'react'

const ProgressBarButtons = (props) => {
    return(
        props.progressBarButtons.map((progressBar, index) =>
            <option key={index} value={index}>ProgressBar {index+1}</option>
        )
    )
}

export default ProgressBarButtons