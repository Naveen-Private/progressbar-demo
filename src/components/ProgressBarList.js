import React from 'react'
import ProgressBar from './ProgressBar'

const ProgressBarList = (props) => {
    return(
        props.progressBars.map((progressBar, index) =>
            <ProgressBar key={index} value={progressBar} />
        )
    )
}

export default ProgressBarList
