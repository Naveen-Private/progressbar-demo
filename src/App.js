import React, { Component } from 'react'
import styled from 'styled-components'

// Components import
import ProgressBar from './components/ProgressBar'
import Button from './components/Button'

// Styles
import { media } from './assets/appCSS'

const Wraper = styled.div`
  ${props => props.isLoading && `
    position: absolute;
    top: 40%;
    left: 49%;
    border: 4px solid #f3f3f3;
    border-radius: 50%;
    border-top: 4px solid #3498db;
    width: 30px;
    height: 30px;
    -webkit-animation: spin 2s linear infinite; /* Safari */
    animation: spin 2s linear infinite;
    @-webkit-keyframes spin {
      0% { -webkit-transform: rotate(0deg); }
      100% { -webkit-transform: rotate(360deg); }
    }
    
    @keyframes spin {
      0% { transform: rotate(0deg); }
      100% { transform: rotate(360deg); }
    }
  `}
`

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 18px;
`

const Title = styled.div`
  display: ${props => props.display ? 'none' : 'block'};
  font-size: 21px;
  margin-bottom: 18px;
  margin-top: 18px;
`

const ButtonContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  width: 650px;
  margin-top: 36px;
  ${media.mobile`
    flex-direction: column;
    width: 100%;
    margin-top: 18px;
  `}
`

const ButtonWrap = styled.div`
    margin-left: 18px;
    ${media.mobile`
      display: flex;
      justify-content: space-around;
      margin-top: 18px;
      margin-left: 0px;
    `}
`

const DropDown = styled.select`
  display: ${props => props.display ? 'none' : 'block'};
  height: 38px;
  width: 150px;
  padding: 5px;
  background: transparent;
  ${media.mobile`
      margin: auto;
      margin-bottom: 18px;
  `}
`


class App extends Component {
  constructor (props){
    super(props)
    this.state = {
      btnSelected: 0,
      progressBars: [],
      buttonValues: [],
      maxLimit: 0,
      isLoading: true
    }
    this.changeDropDownValue = this.changeDropDownValue.bind(this)
    this.updateProgressBar = this.updateProgressBar.bind(this)
  }

  changeDropDownValue(e){
    e.preventDefault()
    console.log(e.target.value)
    this.setState({
      btnSelected: e.target.value
    })
  }

  updateProgressBar(e, btnValue){
    e.preventDefault()
    let btnSelected = this.state.btnSelected
    let progressBars = this.state.progressBars;

    progressBars[btnSelected] += btnValue;
    if(progressBars[btnSelected] > 0) {
      this.setState({
        progressBars: progressBars
      })
    } else {
      progressBars[btnSelected] = 0;
      this.setState({
        progressBars: progressBars
      })
    }
    
  }

  componentDidMount() {
    fetch('http://pb-api.herokuapp.com/bars')
      .then(response => response.json())
      .then(data => 
        this.setState({
        progressBars: data.bars,
        buttonValues: data.buttons,
        maxLimit: data.limit,
        isLoading: false
      }));
  }

  render() {
    return (
      <Wraper isLoading={this.state.isLoading}>
        <Container isLoading={this.state.isLoading}>
          <Title display={this.state.isLoading}>Progress Bar Demo</Title>
          {
            this.state.progressBars.map((progressBar, index) =>
              <ProgressBar key={index} value={progressBar} />
          )}
          <ButtonContainer>
            <DropDown display={this.state.isLoading} onChange={(e) => this.changeDropDownValue(e)}>
              {
                this.state.progressBars.map((progressBar, index) =>
                <option key={index} value={index}>ProgressBar {index+1}</option>
              )}
            </DropDown>
            <ButtonWrap>
              {
                this.state.buttonValues.map((buttonvalue, index) => 
                <Button key={index} updateProgressBar={(e) => this.updateProgressBar(e, buttonvalue)} btnText={buttonvalue} />
              )}
            </ButtonWrap>
          </ButtonContainer>
        </Container>
      </Wraper>
    );
  }
}

export default App;
