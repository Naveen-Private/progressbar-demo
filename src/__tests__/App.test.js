import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

// Components import
import App from '../App';
import ProgressBarList from '../components/ProgressBarList'
import ProgressBarButtons from '../components/ProgressBarButtons'
import Button from '../components/Button'

const mockData = {"buttons":[14,12,-37,-23],"bars":[13,80,17],"limit":190}

describe('App', () => {
  //Parent component
  it('App should be defined', () => {
    expect(App).toBeDefined();
  });
  it('App should render correctly with mockdata', async () => {
    expect.assertions(1);
    const data = await mockData;
    expect(data).toEqual(mockData);
  });

  // ProgressBarList component rendering in App - children
  it('ProgressBarList component should be defined', () => {
    expect(ProgressBarList).toBeDefined();
  });
  it('ProgressBar should render correctly', () => {
    const tree = shallow(
      <ProgressBarList progressBars={[10, 20, 30]} />
    );
    expect(tree).toMatchSnapshot();
  });

  // ProgressBarButtons component rendering in App - children
  it('ProgressBarButtons should be defined', () => {
    expect(ProgressBarButtons).toBeDefined();
  });
  it('ProgressBarButtons should render correctly', () => {
    const tree = shallow(
      <ProgressBarButtons progressBarButtons={[10, 20, -30]} />
    );
    expect(tree).toMatchSnapshot();
  });

  // Button component rendering in App - children
  it('Button should be defined', () => {
    expect(Button).toBeDefined();
  });
  it('Button should render correctly', () => {
    const tree = shallow(
      <Button onClick='() => {console.log("some function")}'>12</Button>
    );
    expect(tree).toMatchSnapshot();
  });
  
});


