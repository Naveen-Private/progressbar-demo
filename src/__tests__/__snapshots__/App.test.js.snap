// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Button should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    onClick="() => {console.log(\\"some function\\")}"
>
    12
</Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "onClick": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "onClick": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App ProgressBar should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProgressBarList
    progressBars={
        Array [
            10,
            20,
            30,
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Array [
    <ProgressBar
      value={10}
/>,
    <ProgressBar
      value={20}
/>,
    <ProgressBar
      value={30}
/>,
  ],
  Symbol(enzyme.__nodes__): Array [
    Array [
      <ProgressBar
        value={10}
/>,
      <ProgressBar
        value={20}
/>,
      <ProgressBar
        value={30}
/>,
    ],
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`App ProgressBarButtons should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProgressBarButtons
    progressBarButtons={
        Array [
            10,
            20,
            -30,
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Array [
    <option
      value={0}
>
      ProgressBar 
      1
</option>,
    <option
      value={1}
>
      ProgressBar 
      2
</option>,
    <option
      value={2}
>
      ProgressBar 
      3
</option>,
  ],
  Symbol(enzyme.__nodes__): Array [
    Array [
      <option
        value={0}
>
        ProgressBar 
        1
</option>,
      <option
        value={1}
>
        ProgressBar 
        2
</option>,
      <option
        value={2}
>
        ProgressBar 
        3
</option>,
    ],
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
